File: JavaSolutions/P05/CoinChange.java
/* Write a function that, given: an amount of money and an array of coin denominations 
   computes the number of ways to make amount of money with coins of the available denominations.
 */ 

import java.util.*;

public class CoinChange {
    
    public static int numWays(int amount, int[] denoms, int i) {
	if (amount == 0) return 1;
	if (amount < 0 || i == denoms.length) return 0;
	int n = 0;
	int den = denoms[i];
	for (int c = 0; c * den <= amount; c++) {
	    int rem = amount - c * den;
	    n += numWays(rem, denoms, i+1);
	}
	return n;
    }

    public static int numWaysDP(int amount, int[] denoms, int i, int[][] book) {
	if (amount == 0) return 1;
	if (amount < 0 || i == denoms.length) return 0;
	int n = 0;
	int den = denoms[i];
	for (int c = 0; c * den <= amount; c++) {
	    int rem = amount - c * den;
	    if (i < denoms.length-1 && book[rem][i+1] > 0) n += book[rem][i+1];
	    else n += numWaysDP(rem, denoms, i+1, book);
	}
	book[amount][i] = n;
	return n;
    }

    public static void test(int amount, int[] denoms) {
	System.out.println("Changing amount: $" + amount + " using coins " + Arrays.toString(denoms));
	System.out.println("Number of ways: " + numWays(amount, denoms, 0));
	System.out.println("_____________________________________________");
    }

    public static void testDP(int amount, int[] denoms) {
	System.out.println("Changing amount: $" + amount + " using coins " + Arrays.toString(denoms));
	int[][] book = new int[amount+1][denoms.length];
	System.out.println("Number of ways: " + numWaysDP(amount, denoms, 0, book));
	System.out.println("_____________________________________________");
    }

    
    public static void main(String[] args) {
	int[] denoms = new int[] {5, 10, 25, 1};
	test(100, denoms);
	testDP(100, denoms);
    }
}

File: PythonSolutions/P05/CoinChange.py
# Write a function that, given:
# 1. an amount of money
# 2. an array of coin denominations
# computes the number of ways to make amount of money with coins of the available denominations.

cache = {}

def change(amount, denoms, i):
	if (amount == 0):
		return 1
	if (i == len(denoms)):
		return 0
	numWays = 0
	maxUse = amount / denoms[i]
	for j in range(0, maxUse + 1):
		remaining = amount - j * denoms[i]
		if (remaining, i + 1) in cache:
			numWays += cache[(remaining, i + 1)]
		else:	
			numWays += change(remaining, denoms, i + 1)
	cache[(amount, i)] = numWays
	return numWays

def test(amount, denoms):
	cache = {}
	n = change(amount, denoms, 0)
	print "Number of ways to change $%d using %s = %d" % (amount, str(denoms), n)

denoms = [1, 5, 10, 20]
test(0, denoms)
test(1, denoms)
test(2, denoms)
test(5, denoms)
test(6, denoms)
test(10, denoms)
test(11, denoms)
test(20, denoms)
test(1000, denoms)
